### DISCUSSION TITLE
CMSIS DSP and NN micropython wrappers

### DISCUSSION THREAD

**Discussion by jgpeiro** [micropython/micropython#10200] [Ideas]
Add a wrappers to these libraries should not be too hard and will be usefull for lot of users that need todo some signal processing.

**Comment by jimmo** [comment]
Yes, this would be really cool.

It's not that simple, because API design never is. But it would be helpful to know the sort of uses cases you're interested in. i.e. what DSP functions do you want to use? What does your input data look like, etc? e.g. array.array?

It's also worth looking at ulab, to see if parts of ulab can be accelerated using this. (Or whether ulab is just the natural place to implement these sorts of things from Python).

**Reply by jgpeiro** [reply]
I think that with a simple methods wrapper can be enough.

The methods, I think all FilteringFunctions can be added and then configured via preprocesor defines. Yes, the data should be array.array and even the structures can be created on python side via uctypes.

About add it to ulab, I think CMSIS is not related with numpy/scipy or other PYTHON scientific packages, so I think an specific modcmsis will be better.

```c
#include	"py/obj.h"
#include	"py/runtime.h"
#include	"arm_math.h"

STATIC mp_obj_t mod_cmsis_dsp_iir_lattice_q31_filter(mp_obj_t self_in, mp_obj_t pSrc_in, mp_obj_t pDst_in, mp_obj_t blockSize_in)
{
    arm_iir_lattice_instance_q31 *self;
    mp_buffer_info_t pSrc;
    mp_buffer_info_t pDst;
    uint32_t blockSize;

    self = MP_OBJ_TO_PTR(self_in);
    mp_get_buffer_raise(pSrc_in, &pSrc, MP_BUFFER_READ);
    mp_get_buffer_raise(pDst_in, &pDst, MP_BUFFER_WRITE);
    blockSize = mp_obj_get_int(blockSize_in);

    arm_iir_lattice_q31(self, pSrc.buf, pDst.buf, blockSize);

    return mp_const_none;
}
```

**Comment by v923z** [comment]
@jgpeiro To extend @jimmo's comment above, I'd be willing to add C functions in `ulab` in the [utils](https://micropython-ulab.readthedocs.io/en/latest/ulab-utils.html) module.

**Comment by jonnor** [comment]
There are dynamic native modules for IIR filters, FFT etc in https://github.com/emlearn/emlearn-micropython
It does not necessarily use ARM CMSIS, but at least it is much faster than Python :)
